### PR TITLE
Feature: Add ArrayAssembler

### PR DESCRIPTION
**Pull request by gitlog**
Implementation of IteratorAggregate (https://www.php.net/manual/ru/class.iteratoraggregate.php), ArrayAccess (https://www.php.net/manual/ru/class.arrayaccess.php), Countable (https://www.php.net/manual/ru/class.countable.php) interfaces to provide accessing objects as arrays.
When we has generated class with IteratorAssembler - we can to add only one element to this array. To add more elements, must use ArrayAccess interface. With Countable interface will be available element's counting, with IteratorAggregate - iterating array elements in loop.

### CODE REVIEW

File: src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    /**
+     * @param ContextInterface|TypeContext $context
+     *
+     * @throws AssemblerException
+     */
+    public function assemble(ContextInterface $context)
+    {
+        $class = $context->getClass();
+        $properties = $context->getType()->getProperties();
+        $firstProperty = count($properties) ? current($properties) : null;
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php:39]: It is better to skip if the first property does not exist.
Otherwise the class implements the interfaces, but not the methods.

File: src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+        try {
+
+            $interfaceAssembler = new InterfaceAssembler(\IteratorAggregate::class);
+            if ($interfaceAssembler->canAssemble($context)) {
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php:44]: It does not contain the `getIterator()` method.
I'dd suggest to drop that interface hete and use the `IteratorAssembler` directly instead.

File: src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        }
+    }
+
+    /**
+     * @param ClassGenerator $class
+     * @param Property       $firstProperty
+     *
+     * @throws \Laminas\Code\Generator\Exception\InvalidArgumentException
+     */
+    private function implementConstruct(ClassGenerator $class, Property $firstProperty)
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php:79]: It is better to set the properties default to `[]` instead of using the constructor.
Now you will cause an issue if you mix it with the constructor assembler.
This method can be removed IMO

File: src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+            'visibility' => MethodGenerator::VISIBILITY_PUBLIC,
+            'docblock' => DocBlockGeneratorFactory::fromArray([
+                'tags' => [
+                    [
+                        'name' => 'param',
+                        'description' => 'mixed $offset An offset to check for',
+                    ],
+                    [
+                        'name' => 'return',
+                        'description' => 'boolean true on success or false on failure',
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php:117]: you could add the strict type here as well

File: src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+     */
+    private function implementOffsetGet(ClassGenerator $class, Property $firstProperty)
+    {
+        $methodName = 'offsetGet';
+        $class->removeMethod($methodName);
+
+        $class->addMethodFromGenerator(MethodGenerator::fromArray([
+            'name' => $methodName,
+            'parameters' => ['name' => 'offset'],
+            'body' => sprintf('return $this->%s[$offset];', $firstProperty->getName()),
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php:139]: Maybe safer to check if the offset exists first.

File: src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+     */
+    private function implementOffsetSet(ClassGenerator $class, Property $firstProperty)
+    {
+        $methodName = 'offsetSet';
+        $class->removeMethod($methodName);
+
+        $class->addMethodFromGenerator(MethodGenerator::fromArray([
+            'name' => $methodName,
+            'parameters' => [['name' => 'offset'], ['name' => 'value']],
+            'body' => sprintf("if(isset(\$this->%s[\$offset])) {\n\t\$this->%s[] = \$value;\n} else {\n\t\$this->%s[\$offset] = \$value; \n}", $firstProperty->getName(), $firstProperty->getName(), $firstProperty->getName()),
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php:170]: Why did you add the if statement? You can just set the property index to the value right?

File: src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+     */
+    private function implementOffsetUnset(ClassGenerator $class, Property $firstProperty)
+    {
+        $methodName = 'offsetUnset';
+        $class->removeMethod($methodName);
+
+        $class->addMethodFromGenerator(MethodGenerator::fromArray([
+            'name' => $methodName,
+            'parameters' => ['name' => 'offset'],
+            'body' => sprintf('unisset($this->%s[$offset]);', $firstProperty->getName()),
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php:197]: There is no `unisset` method in php

File: src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+        $class->addMethodFromGenerator(MethodGenerator::fromArray([
+            'name' => $methodName,
+            'body' => sprintf('return count($this->%s);', $firstProperty->getName()),
+            'visibility' => MethodGenerator::VISIBILITY_PUBLIC,
+            'docblock' => DocBlockGeneratorFactory::fromArray([
+                'tags' => [
+                    [
+                        'name' => 'return',
+                        'description' => sprintf("%s Return count of elements", $firstProperty->getType()),
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php:229]: Count doesnt return the type, it returns an int.
you can also use this as real return type.

File: src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php (hunk 3 above)
Comment (gitlog) [reply src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php:79]: How to set the properties default to [] ?

File: src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php (hunk 6 above)
Comment (gitlog) [reply src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php]: Fix in next commit 

Out:

```php
    /**
     * @param mixed $offset An offset to check for
     */
    public function offsetSet($offset, $value)
    {
        if(!isset($offset)) {
        	$this->Prop[] = $value;
        } else {
        	$this->Prop[$offset] = $value; 
        }
    }
```
Use:

```php

$obj[] = $val1;

$obj[10] = $val2;

```

File: src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php (hunk 7 above)
Comment (gitlog) [reply src/Phpro/SoapClient/CodeGenerator/Assembler/ArrayAssembler.php]: Fix in next commit

**Comment by veewee** [comment]
Closing because of inactivity. Feel free to finish and reopen.

**Comment by veewee** [comment]
Closing this one. In v3, the type system will be much better.
Normalizing the type doesn't seem something we want to be doing from this package.
